spacy/syntax/nonproj.pyx

```diff
@@ -77,39 +78,42 @@ def decompose(label):
 def is_decorated(label):
     return DELIMITER in label
 
-def count_decorated_labels(gold_tuples):
+def count_decorated_labels(gold_data):
     freqs = {}
-    for raw_text, sents in gold_tuples:
-        for (ids, words, tags, heads, labels, iob), ctnts in sents:
-            proj_heads, deco_labels = projectivize(heads, labels)
+    for raw_text, doc_annot in gold_data:
+        for raw_annot in doc_annot.raw_annots:
+            proj_heads, deco_deps = projectivize(raw_annot.heads, raw_annot.deps)
             # set the label to ROOT for each root dependent
-            deco_labels = ['ROOT' if head == i else deco_labels[i]
+            deco_deps = ['ROOT' if head == i else deco_deps[i]
                            for i, head in enumerate(proj_heads)]
             # count label frequencies
-            for label in deco_labels:
+            for label in deco_deps:
                 if is_decorated(label):
                     freqs[label] = freqs.get(label, 0) + 1
     return freqs
 
 
-def preprocess_training_data(gold_tuples, label_freq_cutoff=30):
+def preprocess_training_data(gold_data, label_freq_cutoff=30):
     preprocessed = []
     freqs = {}
-    for raw_text, sents in gold_tuples:
-        prepro_sents = []
-        for (ids, words, tags, heads, labels, iob), ctnts in sents:
-            proj_heads, deco_labels = projectivize(heads, labels)
+    for raw_text, doc_annot in gold_data:
+        prepro_annots = []
+        for raw_annot in doc_annot.raw_annots:
+            proj_heads, deco_deps = projectivize(raw_annot.heads, raw_annot.deps)
             # set the label to ROOT for each root dependent
-            deco_labels = ['ROOT' if head == i else deco_labels[i]
+            deco_deps = ['ROOT' if head == i else deco_deps[i]
                            for i, head in enumerate(proj_heads)]
             # count label frequencies
             if label_freq_cutoff > 0:
-                for label in deco_labels:
+                for label in deco_deps:
                     if is_decorated(label):
                         freqs[label] = freqs.get(label, 0) + 1
-            prepro_sents.append(
-                ((ids, words, tags, proj_heads, deco_labels, iob), ctnts))
-        preprocessed.append((raw_text, prepro_sents))
+            proj_annot = RawAnnot(ids=raw_annot.ids, words=raw_annot.words, tags=raw_annot.tags,
+                                  heads=proj_heads, deps=deco_deps, ents=raw_annot.ents,
+                                  brackets=raw_annot.brackets)
+            prepro_annots.append(proj_annot)
+        prepro_doc = DocAnnot(raw_annots=prepro_annots, cats=doc_annot.cats)
```

Comment:
I don't like having `cats` as a named keyword here. I think we should be able to support various doc-level annotation without API modifications.

Reply:
with a `dict` ?

Reply:
actually, in this spot, we can just modify the `DocAnnot.raw_annots` in place (if they're not read-only)

Reply:
I'm not a huge fan of dicts, either. They get very confusing/undocumented very quickly and a named class with named properties would be easier to understand and validate. Everything needs to be renamed, but something like `RawAnnot` -> `TokenAnnot` and this would be `DocAnnot` (and `DocAnnot` would be something else more container-sounding).

Reply:
A counter-point to consider is that APIs which require non-basic objects as arguments are pretty miserable. The misery goes up exponentially the deeper the tree of objects you require.

If we're going to ask the user to pass a non-basic object for the annotations to `begin_training()` or `update()`, I'm pretty reluctant to have that object take a non-basic object to instantiate. A third level --- where it takes an object that *also* takes an object --- is out of the question.

Reply:
Maybe I didn't explain that well. I didn't mean for it to be nested, just renamed. I think it's weirder to have a class (`RawAnnot`, potentially renamed to `TokenAnnot`) for token-level annotations and not a similar kind of class for doc-level annotations. Both would be one level deep in the `AnnotationThingy`.

Reply:
We currently have a [convertor function](https://github.com/explosion/spaCy/pull/4543/files#diff-b6b3e43f85a70fd78f0b9e8395857783R454) for `nlp.update()` that transforms a tuple+dict format into the new objects (cf [test_gold.py#L248](https://github.com/explosion/spaCy/blob/533308b0719478c2d533a9f8647580e04238be71/spacy/tests/test_gold.py#L248)) - I guess we'd want something similar for begin_training(), to avoid some of these complexities? (this part should actually be backwards compatible)